### PR TITLE
fix swift 6 warnings - thread safe tokenizer and model config

### PR DESCRIPTION
**Pull request by davidkoski**
- based on #125 with simpler mechanism (NSLock)

### CODE REVIEW

File: Libraries/LLM/Configuration.swift [hunk 1]
```diff
@@ -26,20 +26,21 @@ public enum StringOrNumber: Codable, Equatable, Sendable {
...
 
+    @Sendable
     private static func createLlamaModel(url: URL) throws -> LLMModel {
         let configuration = try JSONDecoder().decode(
             LlamaConfiguration.self, from: Data(contentsOf: url))
         return LlamaModel(configuration)
     }
 
-    private static var creators: [String: (URL) throws -> LLMModel] = [
+    private var creators: [String: @Sendable (URL) throws -> LLMModel] = [
```
Comment (davidkoski) [inline Libraries/LLM/Configuration.swift:43]: This is the key -- we want to mark it as storing thread safe (Sendable) closures.

File: Libraries/LLM/Configuration.swift [hunk 2]
```diff
@@ -89,18 +90,44 @@ public struct ModelType: RawRepresentable, Codable, Sendable {
...
 
-    public static func registerModelType(
-        _ type: String, creator: @escaping (URL) throws -> LLMModel
+    public func registerModelType(
+        _ type: String, creator: @Sendable @escaping (URL) throws -> LLMModel
     ) {
-        creators[type] = creator
+        lock.withLock {
+            creators[type] = creator
+        }
```
Comment (davidkoski) [inline Libraries/LLM/Configuration.swift:98]: by guarding the mutable state with the lock we can mark the type as Sendable (unchecked)

File: Libraries/LLM/Tokenizer.swift [hunk 3]
```diff
@@ -50,12 +50,35 @@ private func updateTokenizerConfig(_ tokenizerConfig: Config) -> Config {
...
+    private let lock = NSLock()
+
+    /// overrides for TokenizerModel/knownTokenizers
+    private var replacementTokenizers = [
+        "InternLM2Tokenizer": "PreTrainedTokenizer",
+        "Qwen2Tokenizer": "PreTrainedTokenizer",
+        "CohereTokenizer": "PreTrainedTokenizer",
+    ]
+
+    public subscript(key: String) -> String? {
```
Comment (davidkoski) [inline Libraries/LLM/Tokenizer.swift:67]: This ends up being source compatible with `replacementTokenizers: [String:String]` but uses an NSLock for thread safety.

File: Tools/LinearModelTraining/LinearModelTraining.swift [hunk 4]
```diff
@@ -7,7 +7,7 @@ import MLXNN
 import MLXOptimizers
 import MLXRandom
 
-#if swift(>=6.0)
+#if swift(>=5.10)
```
Comment (davidkoski) [inline Tools/LinearModelTraining/LinearModelTraining.swift:10]: Fixing some other warnings -- we don't necessarily need swift 6, but not all swift 5 understands this.

**Review by awni — APPROVED** [review]
🙏 thanks!

**Pull request merged**